### PR TITLE
Activity poller refactor

### PR DESCRIPTION
**Pull request by bergundy**
Refactor the activity poller to ensure it only shuts down once there are no more activity starts or cancellations to deliver.

The previous behavior was to shut down the poller on the first source to complete.
There is an outstanding issue that I did not fix where local activities shut down too early, need to follow up on that.

For Core based SDKs, the implication here is that the graceful shutdown timer should be started immediately after shutdown has been initiated instead of after the poller returns shutdown.
The main reason for this requirement is so activity cancellations can still be delivered to Lang while there are inflight activities (local or non-local).

### CODE REVIEW

File: core/src/core_tests/activity_tasks.rs
Comment (bergundy) [inline core/src/core_tests/activity_tasks.rs:956]: Can remove this now.

File: core/src/worker/mod.rs
Comment (bergundy) [inline core/src/worker/mod.rs:78]: Ended up going with these atomics since I didn't want to do another refactor.
Can revisit in a follow up PR.

File: core/src/worker/mod.rs
Comment (bergundy) [inline core/src/worker/mod.rs:585]: I commented this out a while ago and the code has changed significantly since then.
I'll check this again after the recent changes.

File: core/src/worker/activities/activity_heartbeat_manager.rs
Comment (bergundy) [inline core/src/worker/activities/activity_heartbeat_manager.rs:83]: I just moved the constructor from where it was before, there was a duplicate `impl` block.

File: core/src/worker/activities.rs
Comment (bergundy) [inline core/src/worker/activities.rs:305]: TODO: this seems to require a better name now that it also notifies shutdown.
Maybe `wait_drain_and_notify_heartbeat_manager`...

File: core/src/core_tests/activity_tasks.rs
Comment (bergundy) [reply core/src/core_tests/activity_tasks.rs:956]: Just the last poll.

File: core/src/worker/mod.rs
Comment (bergundy) [reply core/src/worker/mod.rs:78]: Local and non-local activities should be aggregated into a single stream instead.

File: core/src/worker/mod.rs
Comment (bergundy) [reply core/src/worker/mod.rs:585]: Checked, we'd have to revisit this test.

File: tests/integ_tests/workflow_tests/local_activities.rs
Comment (Sushisource) [inline tests/integ_tests/workflow_tests/local_activities.rs:261]: `.last().unwrap()` is a bit more idiomatic

But I think this whole thing may not be needed. Just call the manual cancel token right at the end of the workflow function.

File: test-utils/src/lib.rs
Comment (Sushisource) [inline test-utils/src/lib.rs:306]: I don't love how this ended up everywhere, but I don't want to hold things up on getting rid of it either, so I can see if I can do some dedupe that combines options & wf/act registration all in one go so this gets automatically set when needed, but I don't wanna block the PR on it.

**Review by Sushisource — COMMENTED** [review]
Overall I think this makes good sense, just some opportunities to simplify / clean up

File: core/src/abstractions.rs
Comment (bergundy) [reply core/src/abstractions.rs:98]: Do you see a problem with the current implementation or is it just your opinion that mutex is better than 2 atomics?

File: core/src/abstractions.rs
Comment (bergundy) [reply core/src/abstractions.rs:149]: Yeah, I like this, I'll change

File: core/src/abstractions.rs
Comment (Sushisource) [reply core/src/abstractions.rs:98]: > help you avoid the dance where you increment before acquiring.

That's the main thing I think. There's nothing objectively wrong with it - but that bit of code made me squint for a few minutes

File: core/src/abstractions.rs
Comment (bergundy) [reply core/src/abstractions.rs:167]: That's what I tried initially but I got:
```
cannot move out of type `TrackedOwnedMeteredSemPermit`, which implements the `Drop` trait
```

File: core/src/core_tests/activity_tasks.rs
Comment (bergundy) [reply core/src/core_tests/activity_tasks.rs:518]: I forgot to revert this change after I fix some stuff. Thanks.

File: core/src/core_tests/mod.rs
Comment (bergundy) [reply core/src/core_tests/mod.rs:63]: Fixed, forgot to clean this up after I fixed some issues.

File: core/src/test_help/mod.rs
Comment (bergundy) [reply core/src/test_help/mod.rs:902]: Yeah.. I thought about it and just went with this because it was easier and I was in the phase of getting tests to pass.
I'll change, I like the extension trait.

File: core/src/worker/activities.rs
Comment (bergundy) [reply core/src/worker/activities.rs:55]: 👍

File: core/src/worker/activities.rs
Comment (bergundy) [reply core/src/worker/activities.rs:101]: 👍

File: core/src/worker/activities/activity_heartbeat_manager.rs
Comment (bergundy) [reply core/src/worker/activities/activity_heartbeat_manager.rs:220]: I'll add a docstring.
The name seems right to me.

File: core/src/worker/activities/activity_heartbeat_manager.rs
Comment (bergundy) [reply core/src/worker/activities/activity_heartbeat_manager.rs:223]: Fixed the name as you suggested, updated the docstring, and removed the TODO.

File: core/src/worker/mod.rs
Comment (bergundy) [reply core/src/worker/mod.rs:81]: I just didn't want this to drag on longer than it already is.
I agree a stream is the right approach.

File: core/src/worker/mod.rs
Comment (bergundy) [reply core/src/worker/mod.rs:150]: Oops, leftover from some of my experiments. Thanks.

File: core/src/worker/mod.rs
Comment (bergundy) [reply core/src/worker/mod.rs:292]: Oh thanks!

File: core/src/worker/mod.rs
Comment (bergundy) [reply core/src/worker/mod.rs:608]: I'm leaving this TODO open

File: test-utils/src/lib.rs
Comment (bergundy) [reply test-utils/src/lib.rs:306]: Agree, don't love it either but I didn't want this PR to get out of hand.

File: tests/integ_tests/workflow_tests/local_activities.rs
Comment (bergundy) [reply tests/integ_tests/workflow_tests/local_activities.rs:261]: Changed to use `last` and got rid of the `is_empty` check.

File: core/src/worker/activities.rs
Comment (bergundy) [reply core/src/worker/activities.rs:117]: 👍

File: core/src/worker/activities.rs
Comment (bergundy) [reply core/src/worker/activities.rs:340]: I'm fine with this suggestion 🤷

File: core/src/worker/activities.rs
Comment (bergundy) [reply core/src/worker/activities.rs:305]: I'm just worried about races if we do this synchronously.
I would agree that your approach is better if we had a single place that manages all of the state but that's not the case.

Comment (bergundy) [reply core/src/worker/activities.rs:305]: Another option is to feed this future into the heartbeat manager.

File: core/src/worker/mod.rs
Comment (Sushisource) [reply core/src/worker/mod.rs:608]: I think we can fix this pretty easily by just making sure the mock gets called 1 or more times (which is as easy as adding `.times(1..)`) and then also setting a timeout on the poll. 

So, poll for a bit, see mock returned default, abort polling, assert slots are unchanged.

File: core/src/worker/activities.rs
Comment (bergundy) [reply core/src/worker/activities.rs:305]: I think I managed to get rid of this by using `take_until` on the cancelled stream.

File: core/src/worker/mod.rs
Comment (Sushisource) [inline core/src/worker/mod.rs:512]: This is covering the situation where WFT pollers dying is the reason for shutdown

File: core/src/worker/mod.rs
Comment (Sushisource) [reply core/src/worker/mod.rs:604]: Would still like to fix this before merge

**Review by Sushisource — APPROVED** [review]
I think this looks good! I would like to address the commented out test, but other than that I'm fine with merging. I'll probably take a pass through it after

File: core/src/worker/mod.rs
Comment (bergundy) [reply core/src/worker/mod.rs:604]: Hmm.. but timeouts do eat up permits now.
We don't return anymore, we keep polling in a loop until we get a non-timeout response.
I can change the test to verify that we keep the permit alive as long as the poller times out.

Comment (bergundy) [reply core/src/worker/mod.rs:604]: Changed.

**Pull request merged**
